### PR TITLE
Add reviewers selection to new pull request fixes #26289

### LINKED ISSUE #26289: Assign reviewers on pull request creation
### Feature Description

The pull request creation dialog currently misses the reviewers slection. You can only select an assinngee while creating a pull request. After creating the pull request and opening it you can select the reviewer for it. GitHub allows you to assign the pull request diretly to a reviewer while creating it which is pretty neat. It makes the workflow of suggesting changes for a repo more convenient.

### Screenshots

GitHub:
GitHub

Gitea:
Gitea

### PR DESCRIPTION
**Pull request by CalK16**
The UI looks just like the assignee selection here because I think the actions from the pull request edit dialog don't make much sense here. The assignee selection on the pr create and edit pages also differ.

图片

图片

----

Hi, I've not done this before, so bear with me. I tried to make changes on @splitt3r 's [pull request](https://github.com/go-gitea/gitea/pull/26596) but was rejected due to a permission issue. Therefore, I had to fork my own repository and incorporate what was done in that pull request. In addition to @splitt3r and @sebastian-sauer's commits, I've also added a commit to resolve the conflicts. Furthermore, I have included two other commits to address @jpraet 's comments.

Please let me know if there is anything else I need to address. Thank you!

### CODE REVIEW

**Comment by wxiaoguang** [comment]
Thank you all for the work.

Actually IMO there are still problems in this PR, for example: incorrect CSS classes like `gt-df`.

There is also a blocker in repo-legacy.ts, I have proposed a refactoring PR "Refactor repo legacy #32404", after #32404, I could help to resolve the conflicts here and fix the problems.

**Comment by lunny** [comment]
Does this replace #26596 ?

File: routers/web/repo/issue.go [hunk 1]
```diff
@@ -1217,7 +1259,37 @@ func ValidateRepoMetas(ctx *context.Context, form forms.CreateIssueForm, isPull
 		assigneeIDs = append(assigneeIDs, form.AssigneeID)
 	}
 
-	return labelIDs, assigneeIDs, milestoneID, form.ProjectID
+	// Check reviewers
+	var reviewerIDs []int64
+	if isPull {
+		if len(form.ReviewerIDs) > 0 {
```
Comment (lunny) [inline routers/web/repo/issue.go]:
```suggestion
	if isPull && len(form.ReviewerIDs) > 0 {
```

**Comment by lunny** [comment]
Looks like team reviewers checked by default before I chose anyone.

**Comment by CalK16** [comment]
> Looks like team reviewers checked by default before I chose anyone.
> 
> 

It's because of the `gt-invisible` HTML tag in the template. I've updated that to `tw-invisible`, and I've tested locally, it should be fine now. And I've also modify some `gt-*` classes to `tw-*` according to some previous commits done by other contributors.

File: web_src/js/features/repo-legacy.ts [hunk 2]
```diff
@@ -243,7 +243,8 @@ export function initRepoCommentForm() {
   initListSubmits('select-label', 'labels');
   initListSubmits('select-assignees', 'assignees');
   initListSubmits('select-assignees-modify', 'assignees');
-  initListSubmits('select-reviewers-modify', 'assignees');
+  initListSubmits('select-reviewers', 'reviewers');
+  initListSubmits('select-reviewers-modify', 'reviewers');
```
Comment (wxiaoguang) [inline web_src/js/features/repo-legacy.ts]: What is `select-reviewers-modify`? Why it was changed to `'reviewers'`?

(see below: https://github.com/go-gitea/gitea/pull/32403#discussion_r1826957271)

File: web_src/js/features/repo-legacy.ts [hunk 3]
```diff
@@ -275,6 +276,8 @@ export function initRepoCommentForm() {
         icon = svg('octicon-project', 18, 'tw-mr-2');
       } else if (input_id === '#assignee_id') {
         icon = `<img class="ui avatar image tw-mr-2" alt="avatar" src=${$(this).data('avatar')}>`;
+      } else if (input_id === '#reviewer_id') {
+        icon = `<img class="ui avatar image gt-mr-3" alt="avatar" src=${$(this).data('avatar')}>`;
```
Comment (wxiaoguang) [inline web_src/js/features/repo-legacy.ts]: Still some `gt-xxx` classes?

File: web_src/js/features/repo-issue-sidebar.ts [hunk 4]
```diff
@@ -265,10 +267,12 @@ export function initRepoIssueSidebar() {
   initListSubmits('select-label', 'labels');
   initListSubmits('select-assignees', 'assignees');
   initListSubmits('select-assignees-modify', 'assignees');
+  initListSubmits('select-reviewers', 'reviewers');
   initListSubmits('select-reviewers-modify', 'assignees');
```
Comment (wxiaoguang) [inline web_src/js/features/repo-issue-sidebar.ts]: I have resolved some conflicts. But here I haven't read it carefully.

This PR ever changed it to `initListSubmits('select-reviewers-modify', 'reviewers');`

Was it done intentionally? If yes, what's the reason behind it?

Comment (CalK16) [reply web_src/js/features/repo-issue-sidebar.ts]: Hi, the class for reviewer list has changed to `ui reviewers list` so I was expecting the `initListSubmits` should init that list too. 

https://github.com/CalK16/gitea/blob/941f7fdf806c706af3bcc5e3d4960cb4a0856a1f/templates/repo/issue/new_form.tmpl#L89C17-L89C17

Comment (wxiaoguang) [reply web_src/js/features/repo-issue-sidebar.ts]: I do not think changing `select-reviewers-modify` is right. I do not see any change related to `select-reviewers-modify` in this PR

Comment (CalK16) [reply web_src/js/features/repo-issue-sidebar.ts]: You are probably right. I double checked the `select-reviewers-modify` and found that it only applied to the sidebar for a created pull request. And that is distinguished from the `select-reviewers`.While this is true, I think I've done something wrong based on my previous understanding. Allow me to spend sometime working on it. Thanks for the correction.

Comment (wxiaoguang) [reply web_src/js/features/repo-issue-sidebar.ts]: Let's have this first:  Split issue sidebar into small templates #32444 , to have a clear codebase and then continue.

**Comment by wxiaoguang** [comment]
Made some changes, and merge the duplicate code.

Now we just use one "reviewer_list.tmpl" to handle the selection.

And the new `initIssueSidebarComboList` should be easier to understand and maintain, the legacy code should be refactored to the new mechanism later.

**Review by wxiaoguang — APPROVED** [review]
It's almost done from my side and I would vote my approval.

Some changes seem more than expected, indeed they were done for improving the maintainability, some legacy functions are too complex to maintain.

I have manually tested the New PR page and PR view page. If anything wrong happens, please let me know and I will fix them ASAP.

Screenshots:

<details>

image

image

image

image

</details>

**Review by lunny — APPROVED** [review]

File: services/pull/pull.go [hunk 5]
```diff
@@ -197,7 +210,17 @@ func NewPullRequest(ctx context.Context, repo *repo_model.Repository, issue *iss
 		}
 		notify_service.IssueChangeAssignee(ctx, issue.Poster, issue, assignee, false, assigneeCommentMap[assigneeID])
 	}
-
+	permDoer, err := access_model.GetUserRepoPermission(ctx, repo, issue.Poster)
```
Comment (lunny) [inline services/pull/pull.go:213]: Looks like these changes are duplicated as line 193 `issue_service.ReviewRequestNotify`

**Pull request merged**